### PR TITLE
Walk down internal tree to find DOM node

### PR DESCRIPTION
**Pull request by spicyj**
This reduces our reliance on hierarchical IDs.

### CODE REVIEW

**Comment by spicyj** [comment]
@sebmarkbage This one's for you too. Does adding `_renderedElement` make sense? I did it so that the loop in getNativeComponentFromComposite can be more monomorphic.

**Comment by spicyj** [comment]
^^ Updated to simplify code and remove empty component registry (now unnecessary!).

**Comment by jimfb** [comment]
Looks good to me, I'm assuming this is also destined for 0.15.

**Comment by spicyj** [comment]
Yes.

**Comment by spicyj** [comment]
@sebmarkbage You said you'd prefer to store the rendered node type (native, composite, or empty) on each composite rather than the actual element/node. Like this?

File: src/shared/utils/ReactNodeTypes.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+var invariant = require('invariant');
+
+var ReactNodeTypes = {
+  NATIVE: 0,
+  COMPOSITE: 1,
+  EMPTY: 2,
+
+  getType: function(node) {
+    if (node === null || node === false) {
```
Comment (syranide) [inline src/shared/utils/ReactNodeTypes.js]: Perhaps this logic should be shared for guaranteed consistency? (instantiateReactComponent has it too)

Comment (spicyj) [inline src/shared/utils/ReactNodeTypes.js]: Yeah maybe that should use that.

**Pull request merged**